src/bmffimage.cpp

```diff
@@ -0,0 +1,312 @@
+            result += p[ bBigEndian ? i : (3-i) ];
+        }
+        return result;
+    }
+
+    std::string boxName(uint32_t box)
+    {
+        char           name[5];
+        std::memcpy   (name,&box,4);
+        name[4] = 0   ;
+        return std::string(name) ;
+    }
+
+    static void boxes_check(size_t b, size_t m)
+    {
+        if ( b > m ) {
+#ifdef EXIV2_DEBUG_MESSAGES
+            std::cout << "Exiv2::BmffImage::readMetadata box maximum exceeded" << std::endl;
+#endif
+            throw Error(kerCorruptedMetadata);
+        }
+    }
+
+    bool superBox(uint32_t box)
+    {
+        return      box == moov
+                ||  box == dinf
+                ||  box == iprp
+                ||  box == ipco
+                ||  box == meta
+                ||  box == iinf
+                ||  box == iloc
+        ;
+    }
+    bool fullBox(uint32_t box)
+    {
+        return      box == meta
+                ||  box == iinf
+                ||  box == iloc
+        ;
+    }
+
+    std::string BmffImage::mimeType() const
+    {
+        switch (fileType)
+        {
+            case avif:
+                return "image/avif";
+
+            case heic:
+            case heif:
+                return "image/heif";
+
+            case crx:
+                return "image/x-canon-cr3";
+
+            default:
+                return "image/generic";
+        }
+    }
+
+    void BmffImage::setComment(const std::string& /*comment*/)
+    {
+        // Todo: implement me!
+        throw(Error(kerInvalidSettingForImage, "Image comment", "ISO BMFF"));
+    } // BmffImage::setComment
+
+    void BmffImage::readMetadata()
+    {
+        if (io_->open() != 0)
+        {
+            throw Error(kerDataSourceOpenFailed, io_->path(), strError());
+        }
+        IoCloser closer(*io_);
+        // Ensure that this is the correct image type
+        if (!isBmffType(*io_, false)) {
+            if (io_->error() || io_->eof()) throw Error(kerFailedToReadImageData);
+            throw Error(kerNotAnImage, "ISO BMFF");
+        }
+
+        long              position  = 0;
+        BmffBoxHeader     box       = {0,0};
+        BmffBoxHeader     subBox    = {0,0};
+        size_t            boxes     = 0 ;
+        size_t            boxem     = 1000 ; // boxes max
+
+        while (io_->read((byte*)&box, sizeof(box)) == sizeof(box))
+        {
+            boxes_check(boxes++, boxem);
+            position   = io_->tell();
+            uint32_t length = getLong((byte*)&box.length, bigEndian);
+            uint32_t type   = getLong((byte*)&box.type, bigEndian);
+#ifdef EXIV2_DEBUG_MESSAGES
+            std::cout << "Exiv2::BmffImage::readMetadata: "
+                      << "Position: " << position
+                      << " box type: " << toAscii(type)
+                      << " length: " << length
+                      << std::endl;
+#endif
+
+            if (length == 0) return ;
+
+            if (length == 1)
```

Comment:
Is it possible that this is missing something inside the curly braces?

Reply:
@hassec This code is very much a "boiler" plate image handler.  Although it recognises some file formats, it doesn't parse the file in readMetadata() to locate the Exiv2/IPTC/XMP/ICC data blocks.  The code to do this is in tvisitor.cpp in my book and has to be ported here.  Another matter is to implement writeMetadata() which serialises the _in-memory_ metadata into blocks and updates the file.  I think the necessary code will be of the order of 500 lines.

![exiv2arch](https://user-images.githubusercontent.com/529982/107206861-82f70f00-69f7-11eb-8f38-3e18610f6b73.png)

@1div0 I'm concerned that you haven't updated the PR for a few weeks.  The project schedule for Code Complete is less than 3 weeks away on 2021-02-28.  Are you confident of making the schedule or would you like me to get involved?  https://github.com/Exiv2/exiv2/issues/1018#issuecomment-757979745